lib/matplotlib/image.py

```diff
@@ -274,16 +274,12 @@ def set_alpha(self, alpha):
 
         Parameters
         ----------
-        alpha : float
+        alpha : float or 2-d array or None
         """
-        if alpha is not None and not isinstance(alpha, Number):
-            alpha = np.asarray(alpha)
-            if alpha.ndim != 2:
-                raise TypeError('alpha must be a float, two-dimensional '
-                                'array, or None')
-        self._alpha = alpha
-        self.pchanged()
-        self.stale = True
+        martist.Artist._set_alpha_for_array(self, alpha)
+        if np.ndim(alpha) not in (0, 2):
```

Comment:
Do you not want to check this before calling the setter?

Reply:
No, I think it is better to have the more basic argument-checking from the setter before checking the number of dimensions.

Reply:
Wouldn't that leave `self.alpha` in an inconsistent state?

Reply:
Why would that matter? An exception is being raised either way.  Look at `_ImageBase.set_data`.  All of the validation is being done *after* the attribute assignment:
```python
        self._A = cbook.safe_masked_invalid(A, copy=True)
```
It looks like the `set_data` code would be slightly more efficient and readable if the attribute assignment were done at the end, but that's not the issue here.  For `set_alpha`, both efficiency and logical order of validation checks are better with the ndim check after `set_alpha_for_array`.

Reply:
To resolve this, can we easily write a test that shows how it works?

Reply:
I don't see anything to resolve.  I have already included a test for this case, raising an exception if the wrong number of dimensions is supplied for an image.

Reply:
FWIW, if we already have cases of inconsistent state on invalid input, then I'm slightly less concerned about adding one more.

Reply:
It is impossible to do *all* validation before setting attributes, given independent setters, because valid alpha depends on data, and vice-versa.   That's why one of the tests I added calls `update_scalarmappable`; that's where the validation has to be, because that is where we have both alpha and data and can cross-check them.